File: python/2197.py
from typing import List
from functools import cache


class Solution:
    @cache
    def gcd(self, a: int, b: int) -> int:
        if a < b:  # a >= b
            a, b = b, a
        return b if a % b == 0 else self.gcd(b, a % b)

    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = [nums[0]]
        for num in nums[1:]:
            curr = num
            while stack and self.gcd(curr, stack[-1]) > 1:
                prev = stack.pop()
                curr = curr * prev // self.gcd(curr, prev)
            stack.append(curr)
        return stack


def main():
    print(Solution().replaceNonCoprimes([6, 4, 3, 2, 7, 6, 2]))
    print(Solution().replaceNonCoprimes([2, 2, 1, 1, 3, 3, 3]))
    print(Solution().replaceNonCoprimes([2]))


if __name__ == "__main__":
    main()

File: go/problems/2197.go
package problems

func replaceNonCoprimes(nums []int) []int {
	stack := make([]int, 0, len(nums))
	stack = append(stack, nums[0])

	gdc := func(a, b int) int {
		if b > a {
			a, b = b, a
		}

		for b > 0 {
			a, b = b, a%b
		}

		return a
	}

	for i := 1; i < len(nums); i++ {
		num := nums[i]

		for len(stack) > 0 {
			prev := stack[len(stack)-1]
			g := gdc(num, prev)
			if g == 1 {
				break
			}

			num = prev * num / g
			stack = stack[:len(stack)-1]
		}

		stack = append(stack, num)
	}

	return stack
}
